src/com/google/enterprise/adaptor/filenet/FileNetAdaptor.java

```diff
@@ -91,11 +120,201 @@ Connection getConnection() {
+    @Override
+    public void getDocIds(Checkpoint checkpoint, DocIdPusher pusher)
+        throws IOException, InterruptedException {
+      checkNotNull(checkpoint, "checkpoint may not be null");
+      checkNotNull(pusher, "pusher may not be null");
+      Date timestamp;
+      int id;
+      if (checkpoint.guid == null) {
+        timestamp = new Date();
+        id = 0;
+      } else {
+        timestamp = checkpoint.timestamp;
+        String idStr = checkpoint.guid.toString();
+        id = Integer.parseInt(
+            idStr.substring(idStr.lastIndexOf('-') + 1, idStr.length() - 1));
+      }
+      int maxDocIds = maxFeedUrls - 1;
+      List<Record> records = new ArrayList<>(maxDocIds);
+      for (int i = 0; i < maxDocIds && ++id < 10000; i++) {
+        DocId docid = newDocId(new Id(String.format(idFormat, id)));
+        records.add(new Record.Builder(docid).build());
+      }
+      if (!records.isEmpty()) {
+        Checkpoint newCheckpoint
+           = new Checkpoint(checkpoint.type, timestamp,
+                new Id(String.format(idFormat, id)));
+        records.add(new Record.Builder(newDocId(newCheckpoint))
+            .setCrawlImmediately(true)
+            .build());
+        pusher.pushRecords(records);
+      }
+    }
+
+    @Override
+    public void getDocContent(Id id, Request request, Response response)
+        throws IOException {
+      checkNotNull(id, "id must not be null");
+      checkNotNull(response, "response must not be null");
+      String content = "Hello from document " + id;
+      response.setContentType("text/plain");
+      copyStream(new ByteArrayInputStream(content.getBytes(UTF_8)),
+                 response.getOutputStream());
+    }
+  }
+
+  @VisibleForTesting
+  static class Checkpoint {
+    private static final String ISO_8601 = "yyyy-MM-dd'T'HH:mm:ss.SSSZ";
+    private static final Pattern ZONE_PATTERN =
+        Pattern.compile("(?:(Z)|([+-][0-9]{2})(:)?([0-9]{2})?)$");
+    private static final String ZULU_WITH_COLON = "+00:00";
+    private static final MessageFormat SHORT_FORMAT = new MessageFormat(
+        "type={0}", US);
+    private static final MessageFormat FULL_FORMAT = new MessageFormat(
+        "type={0};timestamp={1,date,yyyy-MM-dd'T'HH:mm:ss.SSSZ};guid={2}", US);
+
+    public String type;
+    public Date timestamp;
+    public Id guid;
+
+    @SuppressWarnings("fallthrough")
+    public Checkpoint(String checkpoint) {
+      checkNotNull(checkpoint, "checkpoint may not be null");
+      logger.info("Checkpoint: '" + checkpoint + "'");
+      Object[] objs;
+      try {
+        if (checkpoint.indexOf(';') < 0) {
+          objs = SHORT_FORMAT.parse(checkpoint);
+        } else {
+          objs = FULL_FORMAT.parse(checkpoint);
+        }
+      } catch (ParseException e) {
+        throw new IllegalArgumentException(
+            "Invalid Checkpoint: " + checkpoint, e);
+      }
+      switch (objs.length) {
+        case 3:
+          timestamp = (Date) objs[1];
```

Comment:
I don't understand why we would parse a Date out of the strings, and then need a special getter to get the only value we ever want. We never want objects, just strings. The v3 Checkpoint took in strings or <Date, Id> pairs, and stored <String, String> pairs. That seems better.